File: 131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string &s,int left,int right){
        while(left<right){
            if(s[left]!=s[right]){
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
    void solve(int idx,string &s,vector<string>&path,vector<vector<string>>&ans){
        if(idx==s.size()){
            ans.push_back(path);
            return;
        }

        for(int end=idx;end<s.size();end++){
            if(isPalindrome(s,idx,end)){
                path.push_back(s.substr(idx,end-idx+1));
                solve(end+1,s,path,ans);
                path.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>>ans;
        vector<string>path;

        solve(0,s,path,ans);

        return ans;
    }
};

File: 131-palindrome-partitioning/palindrome-partitioning.java
class Solution {
    boolean isPalindrome(String s){
        int left=0;
        int right=s.length()-1;

        while(left<=right){
            if(s.charAt(left)!=s.charAt(right))return false;
            left++;
            right--;
        }
        return true;
    }
    void f(String s,List<String>path,List<List<String>>ans,int n){
        if(s.length()==0){
            ans.add(new ArrayList<>(path));
            return;
        }

        for(int idx=0;idx<s.length();idx++){
            String part=s.substring(0,idx+1);

            if(isPalindrome(part)){
                path.add(part);
                f(s.substring(idx+1),path,ans,n);

                path.remove(path.size()-1);
            }
        }
    }
    public List<List<String>> partition(String s) {
        int n=s.length();
        List<List<String>>ans=new ArrayList<>();
        List<String>path=new ArrayList<>();

        f(s,path,ans,n);

        return ans;
    }
}